files/scripts/arp_update

```diff
@@ -89,32 +89,57 @@ while /bin/true; do
+        # in the middle of the below sequence of commands
+        unresolved_kernel_neighbors=$(ip -6 neigh show | grep -v fe80 | grep $vlan | grep -E 'FAILED|INCOMPLETE')
+        failed_kernel_neighbors=$(echo "$unresolved_kernel_neighbors" | grep FAILED | cut -d ' ' -f 1)
+
+        # it's possible for kernel neighbors to fall out of sync with the hardware
+        # this can result in failed neighbors entries that don't have corresponding zero MAC neighbor entries
+        # and therefore don't have tunnel routes installed in the hardware
+        # flush these neighbors from the kernel to force relearning and resync them to the hardware:
+        # 1. for every FAILED or INCOMPLETE neighbor in the kernel, check if there is a corresponding zero MAC neighbor in APPL_DB
+        # 2. if no zero MAC neighbor entry exists, flush the kernel neighbor entry
+        #     - generates the command 'ip neigh flush <neighbor IPv6>' for all such neighbors
+        unsync_neighbors=$(echo "$unresolved_kernel_neighbors" | cut -d ' ' -f 1 | xargs -I{} bash -c "if [[ -z \"\$(sonic-db-cli APPL_DB hget NEIGH_TABLE:$vlan:{} neigh)\" ]]; then echo '{}'; fi")
+        if [[ ! -z "$unsync_neighbors" ]]; then
+            ip_neigh_flush_cmd="echo \"$unsync_neighbors\" | sed -e 's/^/ip neigh flush /' -e 's/$/;/'"
+            eval `eval "$ip_neigh_flush_cmd"`
+        fi
+
+        # generates the following command for each FAILED or INCOMPLETE IPv6 neighbor
         # timeout 0.2 ping <neighbor IPv6> -n -q -i 0 -c 1 -W 1 -I <VLAN name> >/dev/null
-        ping6_template="sed -e 's/^/timeout 0.2 ping /' -e 's/,/ -n -q -i 0 -c 1 -W 1 -I /' -e 's/$/ >\/dev\/null;/'"
-        failed_ip6_neigh_cmd="ip -6 neigh show | grep -v fe80 | grep $vlan | grep -E 'FAILED|INCOMPLETE' | cut -d ' ' -f 1,3 --output-delimiter=',' | $ping6_template"
-        eval `eval $failed_ip6_neigh_cmd`
+        if [[ ! -z "$unresolved_kernel_neighbors" ]]; then
+            ping6_template="sed -e 's/^/timeout 0.2 ping /' -e 's/,/ -n -q -i 0 -c 1 -W 1 -I /' -e 's/$/ >\/dev\/null;/'"
+            failed_ip6_neigh_cmd="echo \"$unresolved_kernel_neighbors\" | cut -d ' ' -f 1,3 --output-delimiter=',' | $ping6_template"
+            eval `eval "$failed_ip6_neigh_cmd"`
+        fi
+
+        # wait for any transient INCOMPLETE neighbors to transition to FAILED
+        # once these transition to FAILED, we can guarantee that the kernel has generated a netlink message for the neighbor
+        # and proceed setting it to permanent INCOMPLETE
+        # this is necessary for:
+        # 1. unsyncronized neighbors that were flushed, then pinged
+        # 2. existing FAILED neighbors which will be set to transient INCOMPLETE by the above ping
+        wait_neighbors="$unsync_neighbors"$'\n'"$failed_kernel_neighbors"
+        if [[ ! -z "$wait_neighbors" ]]; then
+            trans_incomplete_check=$(echo "$wait_neighbors" | sed  -e 's/^/ip neigh show /' -e 's/$/ | grep -q 'INCOMPLETE' \&\&/')
```

Comment:
this will be tricky, there could be packets destined to such neighbors and entries may stay in INCOMPLETE. May be we must just wait for few seconds and continue after the above ping command.